wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,363 @@
 
     Test description:
-    1. Open chrome
-    2. Navigate to the speedometer website - http://browserbench.org/Speedometer/
-    3. Execute the benchmark
+    1. Host a local copy of the Speedometer website, and make it visible to the device via ADB.
+    2. Open chrome via an intent to access the local copy.
+    3. Execute the benchmark - the copy has been modified to trigger the start of the benchmark.
+    4. The benchmark will write to the browser's sandboxed local storage to signal the benchmark
+       has completed. This local storage is monitored by this workload.
+
+    Known working chrome version 83.0.4103.106
+
+    To modify the archived speedometer workload:
+
+    1. Run 'git clone https://github.com/WebKit/webkit'
+
+    2. Copy PerformanceTests/Speedometer to a directory called document_root, renaming Speedometer to Speedometer2.0
+
+    3. Modify document_root/Speedometer2.0/index.html:
+      3a. Remove the 'defer' attribute from the <script> tags within the <head> section.
+      3b. Add '<script>startTest();</script>' to the very end of the <body> section.
+
+    4. Modify document_root/Speedometer2.0/resources/main.js:
+      4a. Add the listed code after this line:
+
+            document.getElementById('result-number').textContent = results.formattedMean;
+
+          Code to add:
 
-    known working chrome version 80.0.3987.149
-    '''
+			if (location.search.length > 1) {
+				var parts = location.search.substring(1).split('&');
+				for (var i = 0; i < parts.length; i++) {
+					var keyValue = parts[i].split('=');
+					var key = keyValue[0];
+					var value = keyValue[1];
+					if (key === "reportEndId") {
+						window.localStorage.setItem('reportEndId', value);
+					}
+				}
+			}
+
+    5. Run 'tar -cpzf speedometer_archive.tgz document_root'
+
+    6. Copy the tarball into the workloads/speedometer directory
+
+    7. If appropriate, update the commit info in the LICENSE file.
+    """
+    supported_platforms = ["android"]
+
+    package_names = ["org.chromium.chrome", "com.android.chrome"]
+    regex = re.compile('text="(\d+.\d+)" resource-id="result-number"')
 
     parameters = [
-        Parameter('speedometer_version', allowed_values=versions, kind=str, default='2.0',
-                  description='''
-                  The speedometer version to be used.
-                  ''')
+        Parameter(
+            "chrome_package",
+            allowed_values=package_names,
+            kind=str,
+            default="com.android.chrome",
+            description="""
+                  The app package for the browser that will be launched.
+                  """,
+        ),
     ]
 
-    requires_network = True
-
     def __init__(self, target, **kwargs):
         super(Speedometer, self).__init__(target, **kwargs)
-        self.gui.timeout = 1500
-        self.gui.uiauto_params['version'] = self.speedometer_version
+        self.archive_server = ArchiveServer()
+        if not target.is_rooted:
```

Comment:
This should be done inside `initialize()` not `__init__` -- do not rely on target connection existing when the workload object is instantiated.